cpp/arcticdb/async/async_store.hpp

```diff
@@ -75,278 +76,266 @@ class AsyncStore : public Store {
-            }, async::TaskScheduler::instance()->io_thread_count() * 2);
+std::vector<folly::Future<bool>> batch_key_exists(
+        const std::vector<entity::VariantKey> &keys) override {
+    std::vector<folly::Future<bool>> res;
+    res.reserve(keys.size());
+    for (auto itr = keys.cbegin(); itr < keys.cend(); ++itr) {
+        res.push_back(async::submit_io_task(KeyExistsTask(itr, library_)));
     }
+    return res;
+}
 
-    std::vector<folly::Future<bool>> batch_key_exists(
-            const std::vector<entity::VariantKey> &keys) override {
-        std::vector<folly::Future<bool>> res;
-        res.reserve(keys.size());
-        for (auto itr = keys.cbegin(); itr < keys.cend(); ++itr) {
-            res.push_back(async::submit_io_task(KeyExistsTask(itr, library_)));
-        }
-        return res;
-    }
 
-    folly::Future<std::vector<VariantKey>> batch_write(
-        std::vector<std::pair<PartialKey, SegmentInMemory>> &&key_seg_pairs,
-        const std::shared_ptr<DeDupMap> &de_dup_map,
-        const BatchWriteArgs &args) override {
-        auto key_segments = std::move(key_seg_pairs);
-        std::vector<folly::Future<VariantKey>> futs;
-        futs.reserve(key_seg_pairs.size());
-        std::size_t write_count = args.lib_write_count == 0 ? 16ULL : args.lib_write_count;
-
-        auto encode_futs = folly::window(key_segments, [*this](auto &&ks) {
-            auto [key, seg] = std::forward<decltype(ks)>(ks);
-            return async::submit_cpu_task(
-                EncodeAtomTask(std::move(key),
-                               ClockType::nanos_since_epoch(),
-                               std::move(seg),
-                               codec_,
-                               encoding_version_));
-        }, write_count);
-
-        for (folly::Future<storage::KeySegmentPair>& encode_fut : encode_futs) {
-            futs.emplace_back(
-                std::move(encode_fut).thenValue([de_dup_map](auto &&ks) -> std::pair<VariantKey, std::optional<Segment>> {
-                        auto key_seg = std::forward<decltype(ks)>(ks);
-                        return lookup_match_in_dedup_map(de_dup_map, std::move(key_seg));
-                    })
-                    .via(&async::io_executor()).thenValue([lib = library_](auto &&item) {
-                        auto key_opt_segment = std::forward<decltype(item)>(item);
-                        if (key_opt_segment.second)
-                            lib->write(Composite<storage::KeySegmentPair>({VariantKey{key_opt_segment.first},
-                                                                           std::move(*key_opt_segment.second)}));
-
-                        return key_opt_segment.first;
-                    }));
-        }
+    folly::Future<SliceAndKey> async_write(
+            folly::Future<std::tuple<PartialKey, SegmentInMemory, pipelines::FrameSlice>> &&input_fut,
```

Comment:
A struct would avoid this tuple being redefined everywhere this is called from.

Reply:
I'm not 100% sure I like it were we do that elsewhere, it makes it more opaque when it's not used frequently enough for people to know what it is. At the moment this is only called from once place, if it starts to get more widely used maybe we can consider doing it then.